### PR TITLE
feature/GEODE-5505 Cache listener not invoked on a retried destroy() operation

### PR DESCRIPTION
**Pull request by bschuchardt**
This PR has two commits.  The first reverts the previous attempt to fix GEODE-5505.  Those changes only addressed destroy() operations but the problem could occur with any operation.

The second commit implements a delay in region synchronization operations to allow clients to retry their operations.  Region synchronization occurs when another server hosting the same region crashes.  It uses a RegionVersionVector to comb through other servers to find partially distributed operations and send the associated entries.

Without the delay it is possible for the sync to complete before a client retries its operation, so that the client's replay encounters the state change it already made in its previous attempt.

With this change a client's replay will either be on a server that didn't see the previous attempt, or it will be on a server that did see the previous attempt & has the operation recorded in its EventTracker.  In the first case the operation will be performed again and transmitted to other servers.  In the second case the server will not apply the operation to its cache but will forward it to other servers.

The delay is based on the server's "maximum time between pings interval", which defaults to 1 minute.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java [hunk 1]
```diff
@@ -381,8 +381,8 @@ public boolean virtualPut(EntryEventImpl event, boolean ifNew, boolean ifOld,
         // if the listeners have already seen this event, then it has already
         // been successfully applied to the cache. Distributed messages and
         // return
-        if (logger.isDebugEnabled()) {
-          logger.debug("DR.virtualPut: this cache has already seen this event {}", event);
+        if (isTraceEnabled) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java]: did you want to revert this change?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/Put61.java [hunk 2]
```diff
@@ -235,7 +235,7 @@ public void cmdExecute(final Message clientMessage, final ServerConnection serve
     } catch (InvalidDeltaException ide) {
       logger.info(LocalizedMessage.create(
           LocalizedStrings.UpdateOperation_ERROR_APPLYING_DELTA_FOR_KEY_0_OF_REGION_1,
-          new Object[] {key, regionName, ide.getMessage()}));
+          new Object[] {key, regionName}));
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/Put61.java]: did you want to revert this change?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/Put65.java [hunk 3]
```diff
@@ -420,7 +420,7 @@ public void cmdExecute(final Message clientMessage, final ServerConnection serve
     } catch (InvalidDeltaException ide) {
       logger.info(LocalizedMessage.create(
           LocalizedStrings.UpdateOperation_ERROR_APPLYING_DELTA_FOR_KEY_0_OF_REGION_1,
-          new Object[] {key, regionName, ide.getMessage()}));
+          new Object[] {key, regionName}));
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/Put65.java]: did you want to revert this change?

File: geode-core/src/main/java/org/apache/geode/internal/i18n/LocalizedStrings.java [hunk 4]
```diff
@@ -6236,7 +6236,7 @@
   public static final StringId RegionAttributesCreation__CLONING_ENABLE_IS_NOT_THE_SAME_THIS_0_OTHER_1 =
       new StringId(4741, "Cloning enabled is not the same: this:  {0}  other:  {1}");
   public static final StringId UpdateOperation_ERROR_APPLYING_DELTA_FOR_KEY_0_OF_REGION_1 =
-      new StringId(4742, "Error applying delta for key {0} of region {1}: {2}");
+      new StringId(4742, "Error applying delta for key {0} of region {1}");
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/i18n/LocalizedStrings.java]: did you want to revert this change?

Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/i18n/LocalizedStrings.java]: Oh, yeah - thanks for noticing that Darrel.  I'll put those logging changes back in place when I merge to develop.  I don't think they warrant another precheckin run.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java [hunk 5]
```diff
@@ -256,11 +258,23 @@ public void syncForCrashedMember(final InternalDistributedMember id, final Profi
...
-      // bug #49601 - don't synchronize until GII has been performed
-      public void run() {
+    // schedule the synchronization for execution in the future based on the client health monitor
+    // interval. This allows client caches to retry an operation that might otherwise be recovered
+    // through the sync operation. Without associated event information this could cause the
+    // retried operation to be mishandled. See GEODE-5505
+    long delay;
+    try {
+      delay = dr.getGemFireCache().getCacheServers().stream().max(
+          (o1, o2) -> o1.getMaximumTimeBetweenPings() - o2.getMaximumTimeBetweenPings()).get()
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: I'm new to the Stream API but I find this code a bit obscure. I also don't like the need to catch the exception. The three call of getMaximumTimeBetweenPings makes it very long and the third call is hard to understand initially.
After fumbling around for a while in my IDE I came up with this which I think is easier to understand. It compiles; I'm not sure it does all the right things at runtime:
    final long delay = dr.getGemFireCache().getCacheServers().stream().mapToLong(o -> o.getMaximumTimeBetweenPings()).max().orElse(0L);

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/PRBucketSynchronizationDUnitTest.java [hunk 6]
```diff
@@ -211,44 +210,22 @@ private void verifySynchronized(VM vm, final InternalDistributedMember crashedMe
...
-          if (!re.isTombstone()) {
-            if (!dumped) {
-              dumped = true;
-              bucket.dumpBackingMap();
-            }
-            waitingFor = "Object5 is not a tombstone but should be: " + re;
-            return false;
-          }
-          return true;
+      Awaitility.await().until(() -> {
```
Comment (dschneider-pivotal) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/PRBucketSynchronizationDUnitTest.java]: Are these changes needed because of the change to using a system timer for scheduling the sync? If so do these tests now take longer to run as they wait on the system timer?

Comment (bschuchardt) [reply geode-core/src/distributedTest/java/org/apache/geode/cache30/PRBucketSynchronizationDUnitTest.java]: These changes were just to replace a WaitCriterion with Awaitility and to remove debug dumping of the RegionMap

File: geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java (hunk 5 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Your solution looks nice.  In my code I'm just using max(Comparator) to sort the cache servers.

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Something that might be worth cleaning up about this code is that it later does some checks and decides that no sync is needed. For example it does this one:
dr.getDataPolicy().withPersistence() && persistentId == null
and then in synchronizeForLostMember() it does nothing if we don't have concurrency checks enabled.
It seems like all these checks could be done synchronously before hand this off to the scheduler.
It also seems like getRegionForDeltaGII could be renamed (i.e. getRegionForLostMemberSynchronization) or turned into another precondition (that our region is a DistributedRegion). Better would be just to call a method on InternalRegion that does nothing on LocalRegion and is only implemented on DIstributedRegion instead of doing this instanceof check.

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Also in the run2() loop we only check dr.isDestroyed if we are not initialized. Shouldn't we check it once more after the loop to make sure we do the destroy check on an initialized region?
